[SMTChecker] Support to Bool variables

leonardoalt opened the pull request:
This PR adds support to Bool program variables including two types of tests: only Bool and mixed Bool/Integer tests.

Commit:
[SMTChecker] Support to Bool variables

chriseth commented on test/libsolidity/SMTChecker.cpp:
```diff
@@ -329,6 +329,104 @@ BOOST_AUTO_TEST_CASE(ways_to_merge_variables)
+					}
+				}
+				bool xor_x_y = (x && !y) || (!x && y);
+				assert(!xor_x_y || a > 0);
```
Why not `xor_x_y == (a > 0)`?

leonardoalt replied:
Works too. My initial goal was to have it as an implication: `xor_x_y -> a > 0`, but your example is stronger.

chriseth commented on test/libsolidity/SMTChecker.cpp:
```diff
@@ -329,6 +329,104 @@ BOOST_AUTO_TEST_CASE(ways_to_merge_variables)
+				uint a;
+				if(x)
+					a = 1;
+				assert(!x || a > 0);
```
Why not `(!x) != (a > 0)`?

leonardoalt replied:
Here the goal was also to have an implication: `x -> a > 0`.

chriseth commented on libsolidity/formal/SMTChecker.cpp:
```diff
@@ -444,7 +444,7 @@ void SMTChecker::arithmeticOperation(BinaryOperation const& _op)
 {
 	solAssert(_op.annotation().commonType, "");
-	if (_op.annotation().commonType->category() == Type::Category::Integer)
+	if (SSAVariable::supportedType(_op.annotation().commonType.get()))
```
Unrelated to this PR, but shouldn't `supportedType` receive its argument via reference?

leonardoalt replied:
Currently it's via pointer, but I can of course change it to reference if that's preferred.

chriseth replied:
The general idea in the codebase is to only use a raw pointer if nullptr is a valid argument.

leonardoalt replied:
I'm changing it back to `Type::Category`, not sure why I went to pointers in the first place...

chriseth commented:
Did you test all bool operators?
 - comparison (less than, ...)
 - equality / inequality
 - negation

More or less related: Did we implement short-circuiting correctly?

chriseth commented on libsolidity/formal/SolverInterface.h:
```diff
@@ -132,10 +133,12 @@ class Expression
 	Expression operator()(Expression _a) const
 	{
 		solAssert(
```
It is probably better to restructure this function to first assert that the arguments are empty, then handle the cases of int and bool function regularly and put a `solAssert(false, ...)` in the final else case.

leonardoalt replied:
Yes, good idea.

chriseth commented on libsolidity/formal/SolverInterface.h:
```diff
@@ -167,9 +170,11 @@ class SolverInterface
 	virtual Expression newFunction(std::string _name, Sort _domain, Sort _codomain)
 	{
-		solAssert(_domain == Sort::Int && _codomain == Sort::Int, "Function sort not supported.");
+		solAssert(_domain == Sort::Int && (_codomain == Sort::Int || _codomain == Sort::Bool), "Function sort not supported.");
```
This could be similarly restructured.

leonardoalt commented:
Negation, equality and inequality yes.

What is the behavior for LT, LE, GT and GE? I couldn't find it in the docs.

> Did we implement short-circuiting correctly?

Where exactly? The encoding has the same structure as the original expression, so I would say yes.

chriseth commented:
`false < true`.

short circuiting: is `if (x || f())` evaluating `f()` in case `x` is true? It should not, i.e. be roughly equivalent to

```
bool result = false;
if (x) { result = true; }
else if (f()) { result = true; }
```

chriseth commented:
I don't think we can properly test this at this stage, though.

chriseth commented:
Perhaps `bool x = false; if (true || (x = true)) {}; assert(!x);` could work.

leonardoalt commented:
I see your point, but I'm not sure this would make any difference for the solver.
This code
```
bool result = false;
if (x) { result = true; }
else if (f()) { result = true; }
```
says that `x -> result` and `f() -> result`. The solver may still just set both as true.

Commit:
[SMTChecker_Bool] Fix PR comments; Add support to gt, ge, lt, le. and tests.

leonardoalt commented:
- Implemented comparison (gt, ge, lt, le) and tests.
- Replaced `Declaration` pointer by reference in `SSAVariable` and `Symbolic*Variable`.
- Applied the suggested restructure for `SolverInterface`.

chriseth commented on libsolidity/formal/SMTChecker.cpp:
```diff
@@ -444,21 +445,36 @@ void SMTChecker::arithmeticOperation(BinaryOperation const& _op)
+		}
+		else // Bool
+		{
+			value = make_shared<smt::Expression>(
```
Please `solAssert` that it is a bool.

chriseth commented on libsolidity/formal/SSAVariable.cpp:
```diff
@@ -26,23 +27,35 @@ using namespace dev;
+	return typeInteger(_category) || typeBool(_category);
+}
+
+bool SSAVariable::typeInteger(Type::Category _category)
```
`isInteger` / `isBool`?

chriseth reviewed (changes requested):
Only tiny changes left. I created a separate issue for the short-circuiting: https://github.com/ethereum/solidity/issues/3711

Commit:
[SMTChecker_Bool] Fix PR review comments: method renaming and solAssert

leonardoalt commented:
Changes applied.

chriseth commented:
There are test failures. Re-running to see if they were infrastructure problems.
